android/lib/src/main/java/com/swmansion/gesturehandler/TapGestureHandler.java

```diff
@@ -60,35 +88,70 @@ private void endTap() {
     } else {
       mHandler.removeCallbacksAndMessages(null);
     }
-    if (++mTapsSoFar == mNumberOfTaps) {
+    if (++mTapsSoFar == mNumberOfTaps && mNumberOfPointers >= mMinNumberOfPointers) {
       activate();
       end();
     } else {
       mHandler.postDelayed(mFailDelayed, mMaxDelayMs);
     }
   }
 
+  private boolean shouldFail() {
+    float dx = mLastX - mStartX + mOffsetX;
+    if (mMaxDeltaX != MAX_VALUE_IGNORE && Math.abs(dx) > mMaxDeltaX) {
+      return true;
+    }
+
+    float dy = mLastY - mStartY + mOffsetY;
+    if (mMaxDeltaY != MAX_VALUE_IGNORE && Math.abs(dy) > mMaxDeltaY) {
+      return true;
+    }
+
+    float dist = dy * dy + dx * dx;
+    return mMaxDistSq != MAX_VALUE_IGNORE && dist > mMaxDistSq;
+  }
+
   @Override
   protected void onHandle(MotionEvent event) {
     int state = getState();
+    int action = event.getActionMasked();
+
+    if (action == MotionEvent.ACTION_POINTER_UP || action == MotionEvent.ACTION_POINTER_DOWN) {
+      mOffsetX += mLastX - mStartX;
+      mOffsetY += mLastY - mStartY;
+      mLastX = GestureUtils.getLastPointerX(event, true);
+      mLastY = GestureUtils.getLastPointerY(event, true);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+      mStartX = mLastX;
+      mStartY = mLastY;
+    } else {
+      mLastX = GestureUtils.getLastPointerX(event, true);
+      mLastY = GestureUtils.getLastPointerY(event, true);
+      mLastEventOffsetX = event.getRawX() - event.getX();
+      mLastEventOffsetY = event.getRawY() - event.getY();
+    }
 
     mLastEventOffsetX = event.getRawX() - event.getX();
     mLastEventOffsetY = event.getRawY() - event.getY();
     mLastX = event.getRawX();
     mLastY = event.getRawY();
+    if (mNumberOfPointers < event.getPointerCount()) {
+      mNumberOfPointers = event.getPointerCount();
+    }
 
-    if (state == STATE_UNDETERMINED) {
-      begin();
-      mTapsSoFar = 0;
+    if (action == MotionEvent.ACTION_DOWN) {
+      if (state == STATE_UNDETERMINED) {
+        mOffsetX = 0;
+        mOffsetY = 0;
+        begin();
+      }
       startTap();
+    } else if (shouldFail()) {
```

Comment:
This seems to assume that if action is ACTION_DOWN handler would never want to fail. But that seems like an incorrect assumption. E.g. wha if we have numberOfTaps=2 and maxDist set to sth small, then we do double tap but the second tap is very far from the first tap?

I mentioned it earlier but I prefer that we structure this by first checking the current state and then deciding what action to take. This would make it easier to verify that it behaves correctly in many edge cases like the one above.

Reply:
I see, I think that failing shoudl be checked in every cases